### DISCUSSION TITLE
Trying to fix my "custom music player" fish script

### DISCUSSION THREAD

**Discussion by walksanatora** [fish-shell/fish-shell#9770] [Q&A]
so i have the function
```sh
    function pl
        set -fx state_file (mktemp)
        function skip_handle -s USR1
            echo "state: $state_file"
        end
        function play_file -a file
            echo "state: $state_file"
            set ext (string split -r -m1 . {$file})[2]
            if test "$ext" = m3u
                for file2 in (cat $file | sort -R)
                    if test (string sub -l 1 $file2) = '@'
                        string sub -s 2 $file2 | xargs play
                    else
                        play_file "$file2"
                    end
                end
            else if test "$ext" = mp3 -o "$ext" = wav
                echo play >$state_file
                play "$file" &
                wait
            else if test "$ext" = xm
                echo xmp >$state_file
                xmp "$file" &
                wait
            end
        end
        set stop 0
        while test "$stop" = 0
            if test -n "$shuf"
                set argv (shuf -e (string collect $argv))
            end
            for file in $argv
                play_file "$file"
                if test -z "$loop"
                    set stop 1
                end
            end
        end
        #cleanup
        rm "$state_file"
        functions --erase skip_handle
        functions --erase play_file
    end
```
whenever I `kill -10 <fish pid running this function>`    it does not print the status: because it is currently playing music (and the `play` command does not with when &'d)
only after the song finishes does it print the `status:`

**Comment by faho** [comment]
Fish does not currently support backgrounding functions. This is #238. If `play` is a function, your best bet is to instead call it as `fish -c 'play $argv' $file &`.

If it isn't, the issue is that nothing here has told it to react to the USR1 signal (that's number 10). What do you expect to react here? Fish has been signaled but it hasn't been told to do anything. So either send the signal to `play`, or add a `--on-signal` function that does something.

Or, alternatively, your actual best bet is to use a program built for the purpose - mpd with ncmpcpp might work for you.

**Reply by walksanatora** [reply]
play is a comand (from sox)

**Reply by walksanatora** [reply]
(and also it may say "m3u" but I added some questionable features to my local m3u files, ~~kinda like how I added the ability for a gopher server to request to run commands on the client~~)

**Reply by faho** [reply]
In that case, either signal the play process, or tell fish to do something on USR1. This isn't something that happens automatically. You need to install a signal handler like

```fish
function usr1 --on-signal usr1
   echo I HAVE BEEN SIGNALLED!
end
```

What that needs to do is up to you.

**Reply by walksanatora** [reply]
> ```shell
>         function skip_handle -s USR1
>             echo "state: $state_file"
>         end
> ```

I allready am telling fish to do something but nothing is happening

**Reply by faho** [reply]
Ah, you're `wait`ing. While that's happening, fish is waiting, and not doing anything else. The signal handler will only be called *after* the `wait` is done.

Like I alluded to above, fish's concurrency facilities (`&` and `wait` and friends) are quite limited. The actual scripts run in a single thread. Signal/event/etc handlers are just checked after every command.

I'm not entirely sure why you are first backgrounding `play` and then just calling `wait` right after?

**Reply by walksanatora** [reply]
because if i tried `play <file>` or `xmp <file>` it was also waiting for that command to exit before handling the signal, if I try with 
```
        function play_file -a file
            echo "state: $state_file"
            set ext (string split -r -m1 . {$file})[2]
            if test "$ext" = m3u
                for file2 in (cat $file | sort -R)
                    if test (string sub -l 1 $file2) = '@'
                        string sub -s 2 $file2 | xargs play_file
                    else
                        play_file "$file2"
                    end
                end
            else if test "$ext" = mp3 -o "$ext" = wav
                echo play >$state_file
                play "$file"
            else if test "$ext" = xm
                echo xmp >$state_file
                xmp "$file"
            end
        end
```
and then `kill -10 fish` nothing happens (no output) untill i ^C the xmp/play command

**Reply by faho** [reply]
Okay, let's step back.

If you do `play foo`, without the `&`, then that means fish starts play and waits for it to exit. No fish signal handler function can run during.

If you do `play foo &`, that backgrounds `play`. Fish script can continue running, signal handlers can run.

But once you do `wait`, you're back to fish waiting for `play` to exit again.

Basically this idiom:

```fish
foo &
wait
```

is almost entirely useless. It's essentially the same as `foo`.

**Reply by walksanatora** [reply]
so then how should I properlly "start this program and wait for it to exit before starting the next loop while trapping signals"

**Reply by faho** [reply]
Currently: You don't. Fish script is entirely single-threaded so while something is running in the foreground signal handlers don't run.

Which is where I will point you towards trying a more capable music player again.

**Reply by walksanatora** [reply]
(and to think the entire reason I am doing this is so the python script I use to controll my music via keyboard keys+ mouse buttons (for when the lid is closed) can skip the tracker music... I could just make it so it is `~/.wmp-controll` is a file)

**Reply by walksanatora** [reply]
i did the stupid ~/.wmp-controll and it works
